File: backend/utils/validators.js
function isNonEmpty(value) {
  if (typeof value !== 'string') return false;
  return value.trim().length > 0;
}

const { isValidIndianMobile } = require('./phoneUtils');

function isEmail(value) {
  return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(value);
}

function validateRegisterInput({ name, email, password, phone }) {
  if (!isNonEmpty(name)) return 'Name is required';
  if (!isEmail(email)) return 'Valid email is required';

  if (!isValidIndianMobile(phone)) {
    return 'Enter a valid 10-digit Indian mobile number (starting with 6–9)';
  }
  
  const strongPasswordRegex = /^(?=.*[A-Z])(?=.*[a-z])(?=.*\d)(?=.*[@$!%*#?&])[A-Za-z\d@$!%*#?&]{8,}$/;
  if (!password || !strongPasswordRegex.test(password)) {
    return 'Password must be at least 8 characters and contain: an uppercase letter, a lowercase letter, a number, and a special character (@$!%*#?&)';
  }
  return null;
}

function validateLoginInput({ email, password }) {
  if (!isEmail(email)) return 'Valid email is required';
  if (!isNonEmpty(password)) return 'Password is required';
  return null;
}

function validateNewsInput({ title_en, content_en, title_hi, content_hi, title_mr, content_mr, categoryIds, source_name, source_url }) {
  if (!isNonEmpty(title_en)) return 'English title is required';
  if (!isNonEmpty(content_en)) return 'English content is required';
  if (!isNonEmpty(title_hi)) return 'Hindi title is required';
  if (!isNonEmpty(content_hi)) return 'Hindi content is required';
  if (!isNonEmpty(title_mr)) return 'Marathi title is required';
  if (!isNonEmpty(content_mr)) return 'Marathi content is required';
  
  if (!categoryIds || !Array.isArray(categoryIds) || categoryIds.length === 0) {
    return 'At least one category must be selected';
  }

  if (source_url && isNonEmpty(source_url)) {
    try {
      new URL(source_url);
    } catch (_) {
      return 'Valid source URL is required';
    }
    if (!isNonEmpty(source_name)) return 'Source name is required when source URL is provided';
  }

  if (source_name && isNonEmpty(source_name)) {
    if (!source_url || !isNonEmpty(source_url)) return 'Source URL is required when source name is provided';
  }

  return null;
}

module.exports = {
  validateRegisterInput,
  validateLoginInput,
  validateNewsInput,
};

File: lib/utils/validators.dart
class Validators {
  Validators._();

  static final RegExp _emailRegex =
      RegExp(r'^[\w\-\.]+@([\w\-]+\.)+[\w\-]{2,4}$');

  static final RegExp passwordPattern = RegExp(
    r'^(?=.*[A-Z])(?=.*[a-z])(?=.*\d)(?=.*[@$!%*#?&])[A-Za-z\d@$!%*#?&]{8,}$',
  );

  static final RegExp otpPattern = RegExp(r'^\d{6}$');

  static final RegExp indianMobilePattern = RegExp(r'^[6-9]\d{9}$');

  static String normalizeIndianPhone(String value) {
    var digits = value.trim().replaceAll(RegExp(r'[\s\-()]'), '');

    if (digits.startsWith('+91')) {
      digits = digits.substring(3);
    } else if (digits.startsWith('91') && digits.length == 12) {
      digits = digits.substring(2);
    } else if (digits.startsWith('0') && digits.length == 11) {
      digits = digits.substring(1);
    }

    return digits;
  }

  static String? email(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Enter your email';
    }
    if (!_emailRegex.hasMatch(value.trim())) {
      return 'Enter a valid email';
    }
    return null;
  }

  static String? password(String? value) {
    if (value == null || value.isEmpty) {
      return 'Enter a password';
    }
    if (!passwordPattern.hasMatch(value)) {
      return 'Password must be at least 8 characters with uppercase, lowercase, number, and special character (@\$!%*#?&)';
    }
    return null;
  }

  static String? confirmPassword(String? value, String password) {
    if (value == null || value.isEmpty) {
      return 'Confirm your password';
    }
    if (value != password) {
      return 'Passwords do not match';
    }
    return null;
  }

  static String? otp(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Enter the verification code';
    }
    if (!otpPattern.hasMatch(value.trim())) {
      return 'OTP must be exactly 6 digits';
    }
    return null;
  }

  static String? indianPhone(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Enter your phone number';
    }
    final normalized = normalizeIndianPhone(value);
    if (!indianMobilePattern.hasMatch(normalized)) {
      return 'Enter a valid 10-digit Indian mobile number (starting with 6–9)';
    }
    return null;
  }
}
